lib/api/ldp/ldp-operation.js

```diff
@@ -0,0 +1,196 @@
+'use strict'
+
+const url = require('url')
+const glob = require('glob')
+const HttpError = require('standard-http-error')
+
+const OPERATIONS = {
+  'head': LdpHeadOperation,
+  'get': LdpGetOperation,
+  'put': LdpPutOperation,
+  'post': LdpPostOperation,
+  'patch': LdpPatchOperation,
+  'delete': LdpDeleteOperation,
+  'copy': LdpCopyOperation,
+  'options': LdpOptionsOperation
+}
+
+class LdpOperation {
+  /**
+   * @param options {object}
+   *
+   * @param options.target {URL} Parsed full request url
+   *
+   * @param options.headers {object}
+   *
+   * @param [options.authentication=null] {Authentication} Authn object,
+   *   contains WebID string, as well as any bearer credentials/tokens. Needed
+   *   for authenticated fetch (of remote group ACLs, of Copy resources, etc).
+   *   Null if request is not authenticated.
+   *
+   * @param [options.bodyStream] {Stream} Request body stream, for parsing when
+   *   needed
+   */
+  constructor (options) {
+    this.authentication = options.authentication
+    this.target = options.target
+    this.headers = options.headers
+    this.bodyStream = options.bodyStream
+  }
+
+  static async from ({req, host}) {
+    const { method, headers, session: { authentication } } = req
+    const target = LdpOperation.parseTargetUrl({req, host})
+    const bodyStream = req
+
+    const Operation = OPERATIONS[method]
+    if (!Operation) {
+      throw new HttpError(405, 'Method not supported')
+    }
+
+    return Operation.from({headers, target, bodyStream, authentication})
+  }
+
+  /**
+   * Taken from utils.getBaseUri()
+   *
+   * @param req {IncomingRequest}
+   * @param host {SolidHost}
+   *
+   * @returns {URL}
+   */
+  static parseTargetUrl ({req, host}) {
+    const protocol = host.serverUri.replace(/:.*/, '')
```

Comment:
@RubenVerborgh re our earlier discussion of whether or not the `host` param is needed - this usage of it while parsing the target is what I was thinking about.

Reply:
The Operation should not probably not be performing this task; the target URL is also needed by the permission checker. (Also relates to our different definitions of “operation” though, where yours includes target but mine doesn't.)

Reply:
I wonder if the name `host` is misleading.. It doesn't refer to the url hostname, it's a host server config object instance. Should I rename it `serverConfig`?

Reply:
Perhaps, yes.

Also node that part of the relevant code and tests are here: https://github.com/solid/node-solid-server/blob/v4.0.14/lib/resource-mapper.js

ResourceMapper currently does two things: re-constructing the intended URL (with host and protocol), and mapping that URL to a file.

Since this first function is shared across all LDP stores, but that second function is unique to the file-based store, it might make sense to move that first code (and associated tests) to here.

Reply:
Sounds good, lets move the relevant (url non-fs) functions in resource-mapper.js to solid-host.js ?

Reply:
Yes!